Sources/Swift/HangTracker.swift

```diff
@@ -0,0 +1,302 @@
+import UIKit
+#endif
+
+// A hang is parameterized by the its minimum duration, HangDuration, and is defined by a period of time lasting longer than HangDuration*2 when the delay
+// of all frames started during that time sums to more than HangDuration. If there is only one frame during this time that is delayed then it is a "fullyBlocking"
+// hang, otherwise it is a "nonFullyBlocking" hang.
+//
+// We measure late frames using a runloop observer. Any frame that is more than 50% delayed has its stacktrace sampled once, when we first detect it is delayed.
+// After the hang ends (or the app restarts in the case of fatal hangs) the most recently recorded stacktrace is used to report the hang. Since a hang is not one
+// stacktrace, like a crash, but rather a period of time, it would make sense to talk about a flamegraph sampled during the hang but the API does not support this.
+// Note that stacktraces are always sampled from a background thread, since it’s the main thread that gets sampled.
+//
+// There are two ways we trigger saving hang data. One is on the main thread, when the hang is over. When each frame finishes rendering we collect its delay and
+// if the sum of delays during the last HangDuration*2 is greather than the threshold, we report the last stack trace as a hang. The other way is on a background
+// thread. Once a frame is delayed more than 50% we sample the thread and start tracking it. If the total delay becomes more than our threshold the event
+// is recorded. If the hang never ends (thus transfering us back to the previous main thread case) then the hang is saved on the filesystem and
+// sent as a fatal hang the next time the app is launched.
+final class HangTracker {
+    
+    static let SentryANRMechanismDataAppHangDuration = "app_hang_duration"
+
+    private let dateProvider: SentryCurrentDateProvider
+    private let threadInspector: ThreadInspector
+    private let debugImageCache: DebugImageCache
+    private let fileManager: SentryFileManager
+    private let crashWrapper: CrashWrapper
+
+    init(
+        dateProvider: SentryCurrentDateProvider,
+        threadInspector: ThreadInspector,
+        debugImageCache: DebugImageCache,
+        fileManager: SentryFileManager,
+        crashWrapper: CrashWrapper,
+        minHangTime: TimeInterval) {
+        self.dateProvider = dateProvider
+        self.threadInspector = threadInspector
+        self.debugImageCache = debugImageCache
+        self.fileManager = fileManager
+        self.crashWrapper = crashWrapper
+        self.lastFrameTime = 0
+        self.minHangTime = minHangTime
+#if canImport(UIKit) && !SENTRY_NO_UIKIT && !os(visionOS) && !os(watchOS)
+        var maxFPS = 60.0
+            if #available(iOS 13.0, tvOS 13.0, *) {
+            let window = UIApplication.shared.connectedScenes.flatMap { ($0 as? UIWindowScene)?.windows ?? [] }.first { $0.isKeyWindow }
+            maxFPS = Double(window?.screen.maximumFramesPerSecond ?? 60)
```

Comment:
I believe the display rate can change dynamically in a few cases like when displaying video. Not seeing any way to subscribe to that change though.